File: android/lib/src/main/java/com/swmansion/gesturehandler/PanGestureHandler.kt
package com.swmansion.gesturehandler

import android.content.Context
import android.view.MotionEvent
import android.view.VelocityTracker
import android.view.ViewConfiguration
import com.swmansion.gesturehandler.GestureUtils.getLastPointerX
import com.swmansion.gesturehandler.GestureUtils.getLastPointerY

class PanGestureHandler(context: Context?) : GestureHandler<PanGestureHandler>() {
  var velocityX = 0f
    private set
  var velocityY = 0f
    private set
  private val defaultMinDistSq: Float
  private var minDistSq = MAX_VALUE_IGNORE
  private var activeOffsetXStart = MIN_VALUE_IGNORE
  private var activeOffsetXEnd = MAX_VALUE_IGNORE
  private var failOffsetXStart = MAX_VALUE_IGNORE
  private var failOffsetXEnd = MIN_VALUE_IGNORE
  private var activeOffsetYStart = MIN_VALUE_IGNORE
  private var activeOffsetYEnd = MAX_VALUE_IGNORE
  private var failOffsetYStart = MAX_VALUE_IGNORE
  private var failOffsetYEnd = MIN_VALUE_IGNORE
  private var minVelocityX = MIN_VALUE_IGNORE
  private var minVelocityY = MIN_VALUE_IGNORE
  private var minVelocitySq = MIN_VALUE_IGNORE
  private var minPointers = DEFAULT_MIN_POINTERS
  private var maxPointers = DEFAULT_MAX_POINTERS
  private var startX = 0f
  private var startY = 0f
  private var offsetX = 0f
  private var offsetY = 0f
  private var lastX = 0f
  private var lastY = 0f
  private var velocityTracker: VelocityTracker? = null
  private var averageTouches = false

  /**
   * On Android when there are multiple pointers on the screen pan gestures most often just consider
   * the last placed pointer. The behaviour on iOS is quite different where the x and y component
   * of the pan pointer is calculated as an average out of all the pointers placed on the screen.
   *
   * This behaviour can be customized on android by setting averageTouches property of the handler
   * object. This could be useful in particular for the usecases when we attach other handlers that
   * recognizes multi-finger gestures such as rotation. In that case when we only rely on the last
   * placed finger it is easier for the gesture handler to trigger when we do a rotation gesture
   * because each finger when treated separately will travel some distance, whereas the average
   * position of all the fingers will remain still while doing a rotation gesture.
   */
  init {
    val vc = ViewConfiguration.get(context)
    val touchSlop = vc.scaledTouchSlop
    defaultMinDistSq = (touchSlop * touchSlop).toFloat()
    minDistSq = defaultMinDistSq
  }

  override fun resetConfig() {
    super.resetConfig()
    minDistSq = MAX_VALUE_IGNORE
    activeOffsetXStart = MIN_VALUE_IGNORE
    activeOffsetXEnd = MAX_VALUE_IGNORE
    failOffsetXStart = MAX_VALUE_IGNORE
    failOffsetXEnd = MIN_VALUE_IGNORE
    activeOffsetYStart = MIN_VALUE_IGNORE
    activeOffsetYEnd = MAX_VALUE_IGNORE
    failOffsetYStart = MAX_VALUE_IGNORE
    failOffsetYEnd = MIN_VALUE_IGNORE
    minVelocityX = MIN_VALUE_IGNORE
    minVelocityY = MIN_VALUE_IGNORE
    minVelocitySq = MIN_VALUE_IGNORE
    minDistSq = defaultMinDistSq
    minPointers = DEFAULT_MIN_POINTERS
    maxPointers = DEFAULT_MAX_POINTERS
    averageTouches = false
  }

  fun setActiveOffsetXStart(activeOffsetXStart: Float): PanGestureHandler {
    this.activeOffsetXStart = activeOffsetXStart
    return this
  }

  fun setActiveOffsetXEnd(activeOffsetXEnd: Float): PanGestureHandler {
    this.activeOffsetXEnd = activeOffsetXEnd
    return this
  }

  fun setFailOffsetXStart(failOffsetXStart: Float): PanGestureHandler {
    this.failOffsetXStart = failOffsetXStart
    return this
  }

  fun setFailOffsetXEnd(failOffsetXEnd: Float): PanGestureHandler {
    this.failOffsetXEnd = failOffsetXEnd
    return this
  }

  fun setActiveOffsetYStart(activeOffsetYStart: Float): PanGestureHandler {
    this.activeOffsetYStart = activeOffsetYStart
    return this
  }

  fun setActiveOffsetYEnd(activeOffsetYEnd: Float): PanGestureHandler {
    this.activeOffsetYEnd = activeOffsetYEnd
    return this
  }

  fun setFailOffsetYStart(failOffsetYStart: Float): PanGestureHandler {
    this.failOffsetYStart = failOffsetYStart
    return this
  }

  fun setFailOffsetYEnd(failOffsetYEnd: Float): PanGestureHandler {
    this.failOffsetYEnd = failOffsetYEnd
    return this
  }

  fun setMinDist(minDist: Float): PanGestureHandler {
    minDistSq = minDist * minDist
    return this
  }

  fun setMinPointers(minPointers: Int): PanGestureHandler {
    this.minPointers = minPointers
    return this
  }

  fun setMaxPointers(maxPointers: Int): PanGestureHandler {
    this.maxPointers = maxPointers
    return this
  }

  fun setAverageTouches(averageTouches: Boolean): PanGestureHandler {
    this.averageTouches = averageTouches
    return this
  }

  /**
   * @param minVelocity in pixels per second
   */
  fun setMinVelocity(minVelocity: Float): PanGestureHandler {
    minVelocitySq = minVelocity * minVelocity
    return this
  }

  fun setMinVelocityX(minVelocityX: Float): PanGestureHandler {
    this.minVelocityX = minVelocityX
    return this
  }

  fun setMinVelocityY(minVelocityY: Float): PanGestureHandler {
    this.minVelocityY = minVelocityY
    return this
  }

  private fun shouldActivate(): Boolean {
    val dx = lastX - startX + offsetX
    if (activeOffsetXStart != MIN_VALUE_IGNORE && dx < activeOffsetXStart) {
      return true
    }
    if (activeOffsetXEnd != MAX_VALUE_IGNORE && dx > activeOffsetXEnd) {
      return true
    }
    val dy = lastY - startY + offsetY
    if (activeOffsetYStart != MIN_VALUE_IGNORE && dy < activeOffsetYStart) {
      return true
    }
    if (activeOffsetYEnd != MAX_VALUE_IGNORE && dy > activeOffsetYEnd) {
      return true
    }
    val distSq = dx * dx + dy * dy
    if (minDistSq != MIN_VALUE_IGNORE && distSq >= minDistSq) {
      return true
    }
    val vx = velocityX
    if (minVelocityX != MIN_VALUE_IGNORE &&
      (minVelocityX < 0 && vx <= minVelocityX || minVelocityX in 0.0f..vx)) {
      return true
    }
    val vy = velocityY
    if (minVelocityY != MIN_VALUE_IGNORE &&
      (minVelocityY < 0 && vx <= minVelocityY || minVelocityY in 0.0f..vx)) {
      return true
    }
    val velocitySq = vx * vx + vy * vy
    return minVelocitySq != MIN_VALUE_IGNORE && velocitySq >= minVelocitySq
  }

  private fun shouldFail(): Boolean {
    val dx = lastX - startX + offsetX
    if (failOffsetXStart != MAX_VALUE_IGNORE && dx < failOffsetXStart) {
      return true
    }
    if (failOffsetXEnd != MIN_VALUE_IGNORE && dx > failOffsetXEnd) {
      return true
    }
    val dy = lastY - startY + offsetY
    if (failOffsetYStart != MAX_VALUE_IGNORE && dy < failOffsetYStart) {
      return true
    }
    return failOffsetYEnd != MIN_VALUE_IGNORE && dy > failOffsetYEnd
  }

  override fun onHandle(event: MotionEvent) {
    val state = state
    val action = event.actionMasked
    if (action == MotionEvent.ACTION_POINTER_UP || action == MotionEvent.ACTION_POINTER_DOWN) {
      // update offset if new pointer gets added or removed
      offsetX += lastX - startX
      offsetY += lastY - startY

      // reset starting point
      lastX = getLastPointerX(event, averageTouches)
      lastY = getLastPointerY(event, averageTouches)
      startX = lastX
      startY = lastY
    } else {
      lastX = getLastPointerX(event, averageTouches)
      lastY = getLastPointerY(event, averageTouches)
    }
    if (state == STATE_UNDETERMINED && event.pointerCount >= minPointers) {
      startX = lastX
      startY = lastY
      offsetX = 0f
      offsetY = 0f
      velocityTracker = VelocityTracker.obtain()
      addVelocityMovement(velocityTracker, event)
      begin()
    } else if (velocityTracker != null) {
      addVelocityMovement(velocityTracker, event)
      velocityTracker!!.computeCurrentVelocity(1000)
      velocityX = velocityTracker!!.xVelocity
      velocityY = velocityTracker!!.yVelocity
    }
    if (action == MotionEvent.ACTION_UP) {
      if (state == STATE_ACTIVE) {
        end()
      } else {
        fail()
      }
    } else if (action == MotionEvent.ACTION_POINTER_DOWN && event.pointerCount > maxPointers) {
      // When new finger is placed down (POINTER_DOWN) we check if MAX_POINTERS is not exceeded
      if (state == STATE_ACTIVE) {
        cancel()
      } else {
        fail()
      }
    } else if (action == MotionEvent.ACTION_POINTER_UP && state == STATE_ACTIVE && event.pointerCount < minPointers) {
      // When finger is lifted up (POINTER_UP) and the number of pointers falls below MIN_POINTERS
      // threshold, we only want to take an action when the handler has already activated. Otherwise
      // we can still expect more fingers to be placed on screen and fulfill MIN_POINTERS criteria.
      fail()
    } else if (state == STATE_BEGAN) {
      if (shouldFail()) {
        fail()
      } else if (shouldActivate()) {
        // reset starting point
        startX = lastX
        startY = lastY
        activate()
      }
    }
  }

  override fun onReset() {
    velocityTracker?.let {
      it.recycle()
      velocityTracker = null
    }
  }

  val translationX: Float
    get() = lastX - startX + offsetX
  val translationY: Float
    get() = lastY - startY + offsetY

  companion object {
    private const val MIN_VALUE_IGNORE = Float.MAX_VALUE
    private const val MAX_VALUE_IGNORE = Float.MIN_VALUE
    private const val DEFAULT_MIN_POINTERS = 1
    private const val DEFAULT_MAX_POINTERS = 10

    /**
     * This method adds movement to {@class VelocityTracker} first resetting offset of the event so
     * that the velocity is calculated based on the absolute position of touch pointers. This is
     * because if the underlying view moves along with the finger using relative x/y coords yields
     * incorrect results.
     */
    private fun addVelocityMovement(tracker: VelocityTracker?, event: MotionEvent) {
      val offsetX = event.rawX - event.x
      val offsetY = event.rawY - event.y
      event.offsetLocation(offsetX, offsetY)
      tracker!!.addMovement(event)
      event.offsetLocation(-offsetX, -offsetY)
    }
  }
}

File: src/web/PanGestureHandler.ts
import { PixelRatio } from 'react-native';
import { State } from '../State';
import { DEFAULT_TOUCH_SLOP } from './constants';
import { EventTypes, GHEvent } from './EventManager';
import GestureHandler from './GestureHandler';

// interface PanConfig extends Config {
//   minDist?: number;
//   maxDist?: number;

//   minPointers?: number;
//   maxPointers?: number;

//   minVelocity?: number;

//   activeOffsetXStart?: number;
//   activeOffsetXEnd?: number;
//   failOffsetXStart?: number;
//   failOffsetXEnd?: number;
//   activeOffsetYStart?: number;
//   activeOffsetYEnd?: number;
//   failOffsetYStart?: number;
//   failOffsetYEnd?: number;
// }
class PanGestureHandler extends GestureHandler {
  readonly DEFAULT_MIN_POINTERS = 1;
  readonly DEFAULT_MAX_POINTERS = 1;

  //
  public velocityX = 0;
  public velocityY = 0;

  private defaultMinDistSq: number = DEFAULT_TOUCH_SLOP * DEFAULT_TOUCH_SLOP;

  private minDistSq = this.defaultMinDistSq;

  private activeOffsetXStart = Number.MAX_SAFE_INTEGER;
  private activeOffsetXEnd = Number.MIN_SAFE_INTEGER;
  private failOffsetXStart = Number.MIN_SAFE_INTEGER;
  private failOffsetXEnd = Number.MAX_SAFE_INTEGER;

  private activeOffsetYStart = Number.MAX_SAFE_INTEGER;
  private activeOffsetYEnd = Number.MIN_SAFE_INTEGER;
  private failOffsetYStart = Number.MIN_SAFE_INTEGER;
  private failOffsetYEnd = Number.MAX_SAFE_INTEGER;

  private minVelocityX = Number.MAX_SAFE_INTEGER;
  private minVelocityY = Number.MAX_SAFE_INTEGER;
  private minVelocitySq = Number.MAX_SAFE_INTEGER;

  private minPointers = 1;
  private maxPointers = 10;

  private startX = 0;
  private startY = 0;
  private offsetX = 0;
  private offsetY = 0;
  private lastX = 0;
  private lastY = 0;

  private lastCachedX = 0;
  private lastCachedTime = 0;

  private activateAfterLongPress = 0;

  //
  public init(ref: number, propsRef: any): void {
    super.init(ref, propsRef);
  }
  //

  get name(): string {
    return 'pan';
  }

  public updateGestureConfig({ ...props }): void {
    super.updateGestureConfig({ enabled: true, ...props });

    for (const key in this.config) {
      if (
        !isNaN(this.config[key]) &&
        this.config[key] !== undefined &&
        this.config[key] !== null
      ) {
        this.hasCustomActivationCriteria = true;
      }
    }

    this.enabled = this.config.enabled as boolean;

    if (this.config.minDist || this.config.minDist === 0) {
      this.minDistSq = this.config.minDist * this.config.minDist;
    }

    if (this.config.minPointers || this.config.minPointers === 0) {
      this.minPointers = this.config.minPointers as number;
    }

    if (this.config.maxPointers || this.config.maxPointers === 0) {
      this.maxPointers = this.config.maxPointers as number;
    }

    if (this.config.minVelocity) {
      //
    }

    if (this.config.shouldCancelWhenOutside) {
      this.setShouldCancelWhenOutside(false);
    }

    if (
      this.config.activeOffsetXStart ||
      this.config.activeOffsetXStart === 0
    ) {
      this.activeOffsetXStart = this.config.activeOffsetXStart as number;
    }

    if (this.config.activeOffsetXEnd || this.config.activeOffsetXEnd === 0) {
      this.activeOffsetXEnd = this.config.activeOffsetXEnd as number;
    }

    if (this.config.failOffsetXStart || this.config.failOffsetXStart === 0) {
      this.failOffsetXStart = this.config.failOffsetXStart as number;
    }

    if (this.config.failOffsetXEnd || this.config.failOffsetXEnd === 0) {
      this.failOffsetXEnd = this.config.failOffsetXEnd as number;
    }

    if (
      this.config.activeOffsetYStart ||
      this.config.activeOffsetYStart === 0
    ) {
      this.activeOffsetYStart = this.config.activeOffsetYStart as number;
    }

    if (this.config.activeOffsetYEnd || this.config.activeOffsetYEnd === 0) {
      this.activeOffsetYEnd = this.config.activeOffsetYEnd as number;
    }

    if (this.config.failOffsetYStart || this.config.failOffsetYStart === 0) {
      this.failOffsetYStart = this.config.failOffsetYStart as number;
    }

    if (this.config.failOffsetYEnd || this.config.failOffsetYEnd === 0) {
      this.activeOffsetYEnd = this.config.activeOffsetYEnd as number;
    }
  }

  protected transformNativeEvent(event: GHEvent): any {
    if (!this.view) return;

    const rect = this.view.getBoundingClientRect();
    const ratio = PixelRatio.get();

    return {
      translationX: this.getTranslationX(),
      translationY: this.getTranslationY(),
      absoluteX: event.x,
      absoluteY: event.y,
      velocityX: this.velocityX * ratio * 10,
      velocityY: this.velocityY * ratio * 10,
      x: event.x - rect.left,
      y: event.y - rect.top,
    };
  }

  protected resetConfig(): void {
    super.resetConfig();

    this.activeOffsetXStart = Number.MAX_SAFE_INTEGER;
    this.activeOffsetXEnd = Number.MIN_SAFE_INTEGER;
    this.failOffsetXStart = Number.MIN_SAFE_INTEGER;
    this.failOffsetXEnd = Number.MAX_SAFE_INTEGER;

    this.activeOffsetYStart = Number.MAX_SAFE_INTEGER;
    this.activeOffsetYEnd = Number.MIN_SAFE_INTEGER;
    this.failOffsetYStart = Number.MIN_SAFE_INTEGER;
    this.failOffsetYEnd = Number.MAX_SAFE_INTEGER;

    this.minVelocityX = Number.MAX_SAFE_INTEGER;
    this.minVelocityY = Number.MAX_SAFE_INTEGER;
    this.minVelocitySq = Number.MAX_SAFE_INTEGER;

    this.minDistSq = this.defaultMinDistSq;

    this.minPointers = this.DEFAULT_MIN_POINTERS;
    this.maxPointers = this.DEFAULT_MAX_POINTERS;

    this.activateAfterLongPress = 0;
  }

  private getTranslationX(): number {
    return this.lastX - this.startX + this.offsetX;
  }
  private getTranslationY(): number {
    return this.lastY - this.startY + this.offsetY;
  }

  //EventsHandling
  protected onDownAction(event: GHEvent): void {
    super.onDownAction(event);
    this.tracker.addToTracker(event);

    if (this.tracker.getTrackedPointersNumber() > 1) {
      event.eventType = EventTypes.POINTER_DOWN;
      this.onPointerAdd(event);
      return;
    }

    this.lastX = this.tracker.getLastAvgX();
    this.lastY = this.tracker.getLastAvgY();

    this.checkUndetermined(event);
    this.checkBegan(event);
  }
  protected onPointerAdd(event: GHEvent): void {
    this.offsetX += this.lastX - this.startX;
    this.offsetY += this.lastY - this.startY;

    this.lastX = this.tracker.getLastAvgX();
    this.lastY = this.tracker.getLastAvgY();

    this.startX = this.lastX;
    this.startY = this.lastY;

    // this.checkUndetermined(event);

    if (this.tracker.getTrackedPointersNumber() > this.maxPointers) {
      if (this.getState() === State.ACTIVE) this.cancel(event);
      else this.fail(event);
    } else this.checkBegan(event);
  }

  protected onUpAction(event: GHEvent): void {
    super.onUpAction(event);

    if (this.tracker.getTrackedPointersNumber() > 1) {
      this.tracker.removeFromTracker(event.pointerId);

      event.eventType = EventTypes.POINTER_UP;
      this.onPointerRemove(event);
      return;
    }

    if (this.getState() === State.ACTIVE) {
      this.lastX = this.tracker.getLastAvgX();
      this.lastY = this.tracker.getLastAvgY();
    }

    this.tracker.removeFromTracker(event.pointerId);

    if (this.getState() === State.ACTIVE) {
      this.end(event);
    } else {
      this.resetProgress();
      this.fail(event);
    }
  }
  protected onPointerRemove(event: GHEvent): void {
    this.offsetX += this.lastX - this.startX;
    this.offsetY += this.lastY - this.startY;

    this.lastX = this.tracker.getLastAvgX();
    this.lastY = this.tracker.getLastAvgY();

    this.startX = this.lastX;
    this.startY = this.lastY;

    if (
      this.getState() === State.ACTIVE &&
      this.tracker.getTrackedPointersNumber() < this.minPointers
    ) {
      this.resetProgress();
      this.fail(event);
      // this.end(event);
    } else this.checkBegan(event);
  }

  protected onMoveAction(event: GHEvent): void {
    this.lastCachedX = this.tracker.getLastAvgX();
    this.tracker.track(event);

    this.lastX = this.tracker.getLastAvgX();
    this.lastY = this.tracker.getLastAvgY();

    this.checkUndetermined(event);
    this.checkBegan(event);

    super.onMoveAction(event);
  }
  protected onOutAction(event: GHEvent): void {
    super.onOutAction(event);
  }
  protected onEnterAction(event: GHEvent): void {
    super.onEnterAction(event);
  }
  protected onCancelAction(event: GHEvent): void {
    super.onCancelAction(event);

    this.reset();
  }
  protected onOutOfBoundsAction(event: GHEvent): void {
    if (this.getShouldCancelWhenOutside()) return;

    this.tracker.track(event);

    this.lastX = this.tracker.getLastAvgX();
    this.lastY = this.tracker.getLastAvgY();

    this.checkBegan(event);

    if (this.getState() === State.ACTIVE) super.onOutOfBoundsAction(event);
  }

  private shouldActivate(): boolean {
    const dx: number = this.getTranslationX();

    if (
      this.activeOffsetXStart !== Number.MAX_SAFE_INTEGER &&
      dx < this.activeOffsetXStart
    ) {
      return true;
    }

    if (
      this.activeOffsetXEnd !== Number.MIN_SAFE_INTEGER &&
      dx > this.activeOffsetXEnd
    ) {
      return true;
    }

    const dy: number = this.getTranslationY();

    if (
      this.activeOffsetYStart !== Number.MAX_SAFE_INTEGER &&
      dy < this.activeOffsetYStart
    ) {
      return true;
    }

    if (
      this.activeOffsetYEnd !== Number.MIN_SAFE_INTEGER &&
      dy > this.activeOffsetYEnd
    ) {
      return true;
    }

    const distanceSq: number = dx * dx + dy * dy;

    if (
      this.minDistSq !== Number.MAX_SAFE_INTEGER &&
      distanceSq >= this.minDistSq
    ) {
      return true;
    }

    const vx: number = this.velocityX;

    if (
      this.minVelocityX !== Number.MAX_SAFE_INTEGER &&
      ((this.minVelocityX < 0 && vx <= this.minVelocityX) ||
        (this.minVelocityX >= 0 && this.minVelocityX <= vx))
    ) {
      return true;
    }

    const vy: number = this.velocityY;
    if (
      this.minVelocityY !== Number.MAX_SAFE_INTEGER &&
      ((this.minVelocityY < 0 && vy <= this.minVelocityY) ||
        (this.minVelocityY >= 0 && this.minVelocityY <= vy))
    ) {
      return true;
    }

    const velocitySq: number = vx * vx + vy * vy;

    return (
      this.minVelocitySq !== Number.MAX_SAFE_INTEGER &&
      velocitySq >= this.minVelocitySq
    );
  }

  private shouldFail(): boolean {
    const dx: number = this.getTranslationX();
    const dy: number = this.getTranslationY();
    const distanceSq = dx * dx + dy * dy;

    if (this.activateAfterLongPress > 0 && distanceSq > this.defaultMinDistSq) {
      return true;
    }

    if (
      this.failOffsetXStart !== Number.MIN_SAFE_INTEGER &&
      dx < this.failOffsetXStart
    ) {
      return true;
    }

    if (
      this.failOffsetXEnd !== Number.MAX_SAFE_INTEGER &&
      dx > this.failOffsetXEnd
    ) {
      return true;
    }

    if (
      this.failOffsetYStart !== Number.MIN_SAFE_INTEGER &&
      dy < this.failOffsetYStart
    ) {
      return true;
    }

    return (
      this.failOffsetYEnd !== Number.MAX_SAFE_INTEGER &&
      dy > this.failOffsetYEnd
    );
  }

  private checkUndetermined(event: GHEvent): void {
    if (
      this.getState() === State.UNDETERMINED &&
      this.tracker.getTrackedPointersNumber() >= this.minPointers
    ) {
      this.resetProgress();
      this.offsetX = 0;
      this.offsetY = 0;
      this.velocityX = 0;
      this.velocityY = 0;

      this.begin(event);

      //Long press
    } else {
      this.velocityX = this.tracker.getVelocityX(event.pointerId);
      this.velocityY = this.tracker.getVelocityY(event.pointerId);
    }
  }

  private checkBegan(event: GHEvent): void {
    if (this.getState() === State.BEGAN) {
      if (this.shouldFail()) this.fail(event);
      else if (this.shouldActivate()) {
        this.activate(event);
      }
    }
  }

  protected activate(event: GHEvent, force = false): void {
    if (this.currentState !== State.ACTIVE) {
      this.resetProgress();
    }

    super.activate(event, force);
  }

  protected resetProgress(): void {
    if (this.getState() === State.ACTIVE) return;

    this.startX = this.lastX;
    this.startY = this.lastY;
  }

  protected onCancel(): void {
    //
  }
  protected onReset(): void {
    //
  }
}

export default PanGestureHandler;
